### ISSUE TITLE
Converting code for Pose detection (like COCO Data to YOLOv8 data)

### ISSUE THREAD

**Issue by yellowjs0304** [ultralytics/ultralytics#1923]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, First of all Thank you for steady upgrade.
It looks great every time, and I tried it every time.

As you know Yolov8(v5) data format has difference with COCO json.
I'm sorry but, is there any script for converting this data? (It looks like Roboflow does Not supporting pose detection format)
I wanna try with my own data(it has COCO pose detection format(Json))

Please help if you have an idea about it.

**Comment by glenn-jocher** [comment]
@yellowjs0304 yes we use https://github.com/ultralytics/JSON2YOLO for converting between COCO JSON and YOLO format.

For Pose I think we have unmerged changes that are required. @AyushExel @Laughing-q do we need a PR to JSON2YOLO to support Pose export?

**Comment by AyushExel** [comment]
Yes currently the converter isn't there in json2yolo. We'll need to build something to port coco annotations to yolo

**Comment by yellowjs0304** [comment]
Then could you please explain the line of in your sample data(coco8-pose)?
I can't find out some page that explain the data. I only guess the data. 

Could you please explain the meaning?

```
# in your coco8-pose>labels>train>000000000036.txt
# class index , center_x, center y, center w, center h(?), keypoint1_x, keypoint_y, visible, keypoint2_x, keypoint2_y,visible, ..., something zero(?)...

0 0.671279 0.617945 0.645759 0.726859 0.519751 0.381250 2.000000 0.550936 0.348438 2.000000 0.488565 0.367188 2.000000 0.642412 0.354687 2.000000 0.488565 0.395313 2.000000 0.738046 0.526563 2.000000 0.446985 0.534375 2.000000 0.846154 0.771875 2.000000 0.442827 0.812500 2.000000 0.925156 0.964063 2.000000 0.507277 0.698438 2.000000 0.702703 0.942187 2.000000 0.555094 0.950000 2.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000

```

**Comment by AyushExel** [comment]
It should be like this Class id, 4 bbox coordinates, 51(17x3) keypoint coordinates
@Laughing-q

**Comment by trungpham2606** [comment]
hi @AyushExel when running the toy dataset you gave, the plotting images during training look so weird:
![image](https://user-images.githubusercontent.com/35657426/231041967-c018978f-c072-4afe-81c8-fe75ea1933f6.png)
please check it.

**Comment by Laughing-q** [comment]
@trungpham2606 please provide the command or the script you used for us to reproduce. :)

**Comment by trungpham2606** [comment]
> @trungpham2606 please provide the command or the script you used for us to reproduce. :)
from ultralytics import YOLO

def train():
    # Load a model
    model = YOLO('yolov8m-pose.yaml')  # build a new model from YAML
    model = YOLO('yolov8m-pose.pt')  # load a pretrained model (recommended for training)
    model = YOLO('yolov8m-pose.yaml').load('yolov8m-pose.pt')  # build from YAML and transfer weights

    # Train the model
    model.train(data='coco8-pose.yaml', epochs=100, imgsz=640)

if __name__ == '__main__':
    train()

the data I used is from the link you provided in yaml file: https://ultralytics.com/assets/coco8-pose.zip

**Comment by Laughing-q** [comment]
@trungpham2606 thanks! I'm able to reproduce this issue. This is because of the shallow copy which does not work with numpy.
https://github.com/ultralytics/ultralytics/blob/95f96dc5bc1ad6fc22fa9ede1c9fe36d4c6abb56/ultralytics/yolo/data/base.py?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L180
I'll fix this soon.

**Comment by trungpham2606** [comment]
> @trungpham2606 thanks! I'm able to reproduce this issue. This is because of the shallow copy which does not work with numpy.
> 
> https://github.com/ultralytics/ultralytics/blob/95f96dc5bc1ad6fc22fa9ede1c9fe36d4c6abb56/ultralytics/yolo/data/base.py?_pjax=%23js-repo-pjax-container%2C%20div%5Bitemtype%3D%22http%3A%2F%2Fschema.org%2FSoftwareSourceCode%22%5D%20main%2C%20%5Bdata-pjax-container%5D#L180
> 
> 
> I'll fix this soon.

I'm doing the training again and everything seems oke. tks for your prompt fix!

**Comment by glenn-jocher** [comment]
@trungpham2606 glad to hear that the issue is resolved now, and thank you for bringing it to our attention. We aim to provide the best quality software to our users and appreciate your input to help us achieve that goal. Please do not hesitate to reach out if you have any further concerns or questions.

**Comment by trungpham2606** [comment]
> @trungpham2606 glad to hear that the issue is resolved now, and thank you for bringing it to our attention. We aim to provide the best quality software to our users and appreciate your input to help us achieve that goal. Please do not hesitate to reach out if you have any further concerns or questions.

Yeah sure (^^). Iam doing some test on my custom dataset, seem I didnt get the expected results. I will create new problem tab about the v8pose tomorrow in case I cant figure out the problem by myself. 
Thank you for your great contribution~

**Comment by glenn-jocher** [comment]
@trungpham2606 thank you for your kind words. We are always here to help you with any issues or concerns you have with Ultralytics YOLOv8, and we appreciate your trust in our software. We look forward to hearing back from you if you encounter any problems, and we are happy to continue supporting you as best we can.

**Comment by ryouchinsa** [comment]
We improved the JSON2YOLO script so that it converts the COCO keypoints format to YOLOv8 format.
https://github.com/ultralytics/JSON2YOLO/issues/45

[RectLabel](https://rectlabel.com) is an offline image annotation tool for object detection and segmentation.
Although this is not an open source program, with RectLabel, you can label polygons and keypoints in the YOLO format.
```
class_index center_x center_y width height x1 y1 v1 x2 v2 y2 x3 y3 v3 ...
0 0.545230 0.616880 0.298794 0.766239 0.522073 0.309332 2 0.540170 0.293193 2 0.499589 0.296503 2 ...
```

A visibility flag v defined as v=0: not labeled, v=1: labeled but not visible, and v=2: labeled and visible.
https://cocodataset.org/#format-data

![yolo_polygon](https://user-images.githubusercontent.com/1954306/209942647-fa9565e0-d4cf-42c2-a382-a8043465765b.jpg)

**Comment by glenn-jocher** [comment]
@ryouchinsa thank you for sharing your experience and the solution. We're glad to hear that you've improved the JSON2YOLO script to convert the COCO keypoint format to the YOLOv8 format. This will help users who want to train their own models using the dataset in this format. Additionally, thank you for introducing RectLabel - an offline image annotation tool that supports labeling polygons and keypoints in YOLOv8 format. It's great to know that there are tools available to make the annotation process easier for YOLOv8 users. Finally, thank you for sharing the sample output format for polygon labels in YOLOv8. The visibility flag is a useful feature for labeling accuracy. We appreciate your contribution to the YOLOv8 community and look forward to more collaborations in the future.

**Comment by ryouchinsa** [comment]
We really appreciate YOLOv5 and YOLOv8 because so many macOS users can train detection/segmentation/pose models and export to the Core ML models easily.
Including the data conversion, we will contribute to your community and look forward to more collaboration in the future.

**Comment by glenn-jocher** [comment]
@ryouchinsa thank you for your kind words and support of the YOLOv8 community. We are glad that our software has been useful to you, and appreciate your contribution to the community. We look forward to further collaborations and working together to improve our software offerings for everyone, including macOS users. Don't hesitate to reach out if you have any further questions or concerns.

**Comment by twilson1111** [comment]
Here is a script and workflow that I used to convert from Roboflow to CVAT and then processed the CVAT output into ultralytics Yolov8 Pose format.

[Here](https://github.com/cvat-ai/cvat/issues/5904#issuecomment-2067068503)

**Comment by glenn-jocher** [comment]
Thanks for sharing your workflow and script for converting data from Roboflow to CVAT, and then to Ultraalytics YOLOv8 Pose format! This resource is really valuable for the community, especially for those looking to streamline their annotation processes across different platforms. 🌟 If anyone has questions or further insights on this workflow, feel free to jump into the discussion!

**Comment by shobhajnilu** [comment]
How to use yolov8 keypoint detection type in iOS project?. After train the Roboflow dataset from google colab, i converted result into .mlpackage and drag & drop into the Xcode project but showing error "Model class has not been generated yet".
how to integrate .mlpackage in swift ios?

**Comment by glenn-jocher** [comment]
For integrating YOLOv8 pose models in iOS, ensure you export with `nms=True` and verify CoreML compatibility. The "Model class not generated" error typically indicates Xcode hasn't processed the model yet - try cleaning build folder and re-adding the .mlpackage. For implementation details, refer to our CoreML deployment guide: https://docs.ultralytics.com/integrations/coreml and Apple's CoreML documentation for Swift examples.
